### ISSUE TITLE
TTD Bookmark Objects/Functionality Unimplemented

### ISSUE THREAD

**Issue by DMHalford** [microsoft/WinDbg-Feedback#170]
### The Bug in Question

Attempting to "Bookmark Current Position" in TTD sessions via the button control in the Timelines window fails (following the "bookmark name" dialog,) with the following error:

    0:000> dx @$cursession.TTD.Bookmarks.AddBookmark("")
    Error: Unable to bind name 'Bookmarks'

This seems to affect the latest version available for download from both the Microsoft Store and the following link: https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/ (v1.2308.2002.0, at the time of writing). It also appears that this issue has been ongoing for quite some time:

> WinDbg Preview does technically have a Bookmark feature so you can keep track of different positions but I also haven’t been able to get it functioning so I’m not going to cover it.

(Such Security, 2020 - https://suchsecurity.com/intro-to-vulnerability-research-debugging-and-intro-to-windbg.html)

### A Partial Solution

I've written a script using the JS API to emulate this feature (see attached). The script modifies the debugger object model to implement the functionality of the "Bookmark Current Position" control, registers a **!ttb** extension command to supply a basic CLI for displaying, adding, removing and renaming bookmarks in a scriptable fashion and additionally ships with a sister DLL extension (compiled only for x64,) to facilitate DML-enhanced output of the current bookmarks list, with clickable 'Time Travel' and deletion links.

[InitBookmarks.zip](https://github.com/microsoftfeedback/WinDbg-Feedback/files/14469838/InitBookmarks.zip)

To use the script, simply extract the Zip file to a directory of your choice and run the following commands:

    .scriptload "<Path-to-script-directory>\InitBookmarks.js"
    !ttb "-h"

### Caveats

- The script does **not** implement the "Timeline Bookmarks" feature of WinDbg alluded to here: https://learn.microsoft.com/en-us/windows-hardware/drivers/debuggercmds/windbg-timeline-preview#timeline-bookmarks. Rather, it makes use of DML-based output as a "next best thing". (If anyone wants to reverse the UI in search of a way to implement bookmark timelines at runtime from a third party perspective, feel free!)
- I do **not** consider myself a "JavaScript developer"; as such, the code is far from elegant and I'm probably duplicating objects all over the place. That said, the code works and is a viable alternative to awaiting an official fix from Microsoft for my personal use cases; hopefully this script will benefit others that would also like to utilise the Bookmarks feature of TTD in the meantime.
- Much of the functionality relies on WinDbg output parsing and, thus, _may_ break with future engine/shell updates. Feedback and bug reports are welcomed, although I do not currently plan on actively maintaining this script (at least in the public domain).
- In its current form, the script loads the sister extension on-demand from the script's own directory. Source code for the sister extension is available on request, although it's a very simple module and its single exported extension command, **!dmlout**, can be duplicated with ease.
- If anyone wants to improve this script (it could probably do with a thorough refactoring), please feel free to do so and distribute the resulting code at your leisure. I would, however, appreciate it if my name remains credited in the comments at the start of the file because, to be honest, I've spent more time on this than I care to admit!

Kudos to Microsoft for the much needed renovation of the WinDbg UI.
An excellent re-work and a much better user experience for us low-level nerds - thank you!

**Comment by KenSykes** [comment]
If you update to the latest windbgnext bookmarks should work.

**Comment by DMHalford** [comment]
Hi Ken, thanks for the heads up.

Can I ask which version you're running?
I've just installed the latest available from both Microsoft Learn (the direct .appinstaller link available here: https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/) and the Windows Store (Client: 1.2407.24003.0 and Engine: 10.0.27668.1000 on an x86-64 system,) and am still faced with the same "Unable to bind name 'Bookmarks'" error.

Thank you.

**Comment by maolike** [comment]
I have the same question.WindDbg  version: 1.2409.17001.0

**Comment by zhuman** [comment]
This is still broken in the latest 1.2502.25002.0  @KenSykes

**Comment by KenSykes** [comment]
Yes, it is. I am tracking down why it is not in the external release.

**Comment by KenSykes** [comment]
Ok so I have a workaround. With the current WinDbg (April 2025 version) the bookmark functionality is in the product but it does not load automatically. However, if you load it manually like below the functionality will work.

I am tracking down the missing pieces to get it to load automatically.

```
0:000> dx @$cursession.TTD.Bookmarks 
Error: Unable to bind name 'Bookmarks' 
0:000> .scriptlist 
Command Loaded Scripts: 
    JavaScript script from 'C:\Program Files\WindowsApps\Microsoft.WinDbg_1.2504.15001.0_x64__8wekyb3d8bbwe\amd64\TTD\Analyzers\HeapAnalysis.js' 
0:000> .scriptload C:\Program Files\WindowsApps\Microsoft.WinDbg_1.2504.15001.0_x64__8wekyb3d8bbwe\amd64\TTD\bookmarks.js 
JavaScript script successfully loaded from 'C:\Program Files\WindowsApps\Microsoft.WinDbg_1.2504.15001.0_x64__8wekyb3d8bbwe\amd64\TTD\bookmarks.js' 
0:000> dx @$cursession.TTD.Bookmarks 
@$cursession.TTD.Bookmarks                 : Bookmark collection 
0:000> dx @$cursession.TTD.Bookmarks.AddBookmark("here") 
@$cursession.TTD.Bookmarks.AddBookmark("here") 
```

**Comment by DMHalford** [comment]
Brilliant stuff! Thanks very much for your work on this @KenSykes - will give your workaround a shot tonight!

**Comment by DMHalford** [comment]
For anyone else following this - I can confirm that the workaround @KenSykes posted above works as expected. Thanks again Ken!

For what it's worth, the same 'bookmarks.js' file referenced in the workaround also seems to function with older versions of WinDbg in case anyone needs to backport for whatever reason. (Personally, I'm not keen on the updated layout of the registers window; simply copying the file to the appropriate directory in an older version of WinDbg gave me the best of both worlds.)

I'll leave this issue open until the next WinDbg release so others can easily find the workaround until bookmarks.js is loaded automatically.
